go-controller/pkg/util/kube.go

```diff
@@ -644,45 +688,123 @@ func GetLbEndpoints(slices []*discovery.EndpointSlice, svcPort kapi.ServicePort,
 			}
 
 			out.Port = *slicePort.Port
-
-			if slice.AddressType == discovery.AddressTypeFQDN {
+			var endpointsPointer *[]discovery.Endpoint
+			switch slice.AddressType {
+			case discovery.AddressTypeFQDN:
 				klog.V(5).Infof("Skipping FQDN slice %s/%s", slice.Namespace, slice.Name)
-			} else { // endpoints are here either IPv4 or IPv6
-				validSlices = append(validSlices, slice)
+				continue
+			case discovery.AddressTypeIPv4:
+				endpointsPointer = &v4EndpointsTmp
+			case discovery.AddressTypeIPv6:
+				endpointsPointer = &v6EndpointsTmp
+			}
+
+			if endpointsPointer == nil {
+				continue
+			}
+
+			for _, ep := range slice.Endpoints {
+				// readiness selection: include endpoints that are either ready
+				// or terminating and serving.
+				if !(includeAllEndpoints || IsEndpointReady(ep) || IsEndpointServing(ep)) {
+					continue
+				}
+
+				// add to v4 or v6 candidate endpoints
+				if len(ep.Addresses) > 0 {
+					*endpointsPointer = append(*endpointsPointer, ep)
+				}
 			}
 		}
 	}
-
-	serviceStr := ""
-	if service != nil {
-		serviceStr = fmt.Sprintf(" for service %s/%s", service.Namespace, service.Name)
+	if len(v4EndpointsTmp) > 0 {
+		out.V4Endpoints = removeDuplicateAddressesFromEndpoints(v4EndpointsTmp)
 	}
-	// separate IPv4 from IPv6 addresses for eligible endpoints
-	for _, endpoint := range getEligibleEndpoints(validSlices, service) {
-		for _, ip := range endpoint.Addresses {
-			if utilnet.IsIPv4String(ip) {
-				klog.V(5).Infof("Adding endpoint IPv4 address %s port %d%s",
-					ip, out.Port, serviceStr)
-				v4IPs.Insert(utilnet.ParseIPSloppy(ip).String())
+	if len(v6EndpointsTmp) > 0 {
+		out.V6Endpoints = removeDuplicateAddressesFromEndpoints(v6EndpointsTmp)
+	}
+
+	klog.V(5).Infof("LB candidate endpoints for %s/%s on port %d are: %v / %v ",
+		slices[0].Namespace, slices[0].Labels[discovery.LabelServiceName], out.Port,
+		out.V4Endpoints, out.V6Endpoints)
+
+	return out
+}
 
-			} else if utilnet.IsIPv6String(ip) {
-				klog.V(5).Infof("Adding endpoint IPv6 address %s port %d%s",
-					ip, out.Port, serviceStr)
-				v6IPs.Insert(utilnet.ParseIPSloppy(ip).String())
+func getEndpointInfoForLogging(ep discovery.Endpoint) string {
+	return fmt.Sprintf("node=%s, addresses=%s, conditions=Ready:%t, Serving:%t, Terminating:%t",
+		*ep.NodeName, ep.Addresses, *ep.Conditions.Ready, *ep.Conditions.Serving,
+		*ep.Conditions.Terminating)
+}
 
+// removeDuplicateAddressesFromEndpoints takes a list of endpoints and removes duplicate IP addresses
```

Comment:
This is also new right as we were not doing this kind of deduplication before? Does this have any relationship with the original issue you are trying to solve or is it a different issue altogether?

Reply:
We were getting away with it by simply keeping a set (from the sets package) of IP addresses (https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/kube.go#L618-L619), since we knew we wanted them all, but now that I'm dealing with endpoints and not with IP addresses, I had to write the logic for it.

I realized I needed it because we have a unit test for it: https://github.com/ovn-org/ovn-kubernetes/blob/5f6cbb16f29239647b8db26d11c15dc33ca27455/go-controller/pkg/util/kube_test.go#L654

Then I looked at previous discussions in upstream kubernetes and saw that duplicate addresses are not removed in kube code and it's left to kube api clients to deal with them: https://github.com/kubernetes/kubernetes/issues/121556

Reply:
The fact that we had a set meant that we dedupped before as well, but what changed is the "how" regardless of the data structure:
```
we keep the IP address on the endpoint with the worse conditions (w.r.t ready, serving, terminating)
and remove the IP from the endpoint with the better conditions
```
Were we doing this with the set before?

Reply:
We deduplicated also before, but without a criterion, since we were just handling IP addresses and not whole endpoints. Mind you, I'm also fine with keeping it simpler.

Reply:
You're right, I didn't get that part either. I'll investigate on Monday.

Reply:
I think that the deduplication is orthogonal to `PublishNotReadyAddresses`, since we want it to happen also when the service is configured to publish all addresses, I'll set it aside.

The point in the comment I linked is that a duplicate endpoint is not common and it's best to consider it as an error case that we resort to only if there's no better alternative. So If the same address appears in two endpoints, one of which is ready and the other of which is serving and terminating, we can be pessimistic about it and consider only the terminating and serving endpoint. We will use it only when we're falling back to terminating and serving endpoints and we will not mix it with ready endpoints. One can also argue that we may as well remove a duplicate endpoint all together if we consider it an error case, but the conversation can go on forever then :) 

As for this little paragraph:

> We already will choose terminating endpoints when no other endpoints exist. We chose at that time to NOT fall back on unready endpoints when no ready endpoints exist. If we want to revisit that, we can, but I don't think it's specific to this case.

I think the choice of words is confusing. I read "unready endpoints" as terminating & non-serving endpoints, otherwise the first sentence would be false and the whole KEP would break :)

Reply:
And of course I'm fine with whatever decision. I had to make a decision in the code and I looked at what is done upstream to justify it.

Reply:
yes, I choose terminating and not serving over terminating and serving over ready.

Reply:
I am just not sure I am making the same interpretation as you of the comment you are linking.

For example
> I don't think that's the common case, though. I think it's more likely that I have N ready and 1 (or small N) that is both ready and unready (dupplicates). In that case, optimism means we include an endpoint in a set which doesn't need it, and risk errors.

I understand here that duplicates should be excluded altogether vs your interpretation that duplicates should be picked through their worst endpoint.
So if i have IP A, B and duplicated C, I would just include A and B, and filter out C.
As in any information about C is un-trusted because it is duplicated.

That is as far as I can read into that comment.

It doesn't offer any hindsight on what should we do, for example, if C was the only IP and it was duplicated. And I don't have a clear picture in my head why the worst condition would be preferred over the best condition in this case.

Reply:
The overall logic might be fine. It is just that the implementation might make people read too much into it. It is more about "if we have a serving & not serving duplicated IP, we just exclude it because we don't know what to do with it; otherwise we include it, it doesn't really matter through which endpoint". If it does matter with what endpoint we include a duplicated IP, then that is the point I am missing.

Reply:
There is also the question of how does this interpretation affect other functionality, like egress services and LB healthcheck.